### ISSUE TITLE
Reverse Engineering Configurations of Neural Text Generation Models

### ISSUE THREAD

**Issue by noowad93** [noowad93/ml-paper-summaries#10]
# 논문 정보 (URL, 저자 등)

[https://arxiv.org/pdf/1903.02134.pdf](https://arxiv.org/pdf/1903.02134.pdf)

Google Research

# 개요

- 본 논문은 **주어진 생성 문장을 보고, 이 문장을 생성한 모델의 설정을 알 수 있을까?**에 대해 조사를 한 논문입니다.
- 생성된 문장에 내포된 특징을 조사함으로써 생성 전략 및 모델링의 약점을 파악할 수도 있고, 실서비스로 사용될 때의 잠재적인 위험성 등을 완화할 수 있을거라 생각합니다.

본 논문의 Contribution을 요약하자면,

- 대규모 생성 모델로부터 생성된 문장의 특징을 분석했습니다.
- 생성된 문장만을 갖고 생성 모델의 세세한 설정 (hyperparameter 등)을 구분할 수 있는 태스크를 제안했습니다.
- (1) 모델링 초이스를 단순 분류기만으로도 알 수 있다는 점 (2) 예측의 난이도가 하이퍼파라미터 설정에 따라 다르다는 점 (3) 단어 순서가 그렇게 중요하지 않다는 점 (syntax나 composition이 아닌 단어 선택에 관련되어 있다는 점) (4) 모델들 사이의 구별은 human-or-machine 예측보다 더 어렵다는 점 등을 알아냈습니다.

# 모델

- 본 연구에서는 Grover를 생성 모델로 이용했습니다.
    - Grover는 Defending against neural fake news. (Zellers et al., 2019)에서 제안된 페이크 뉴스 생성 모델 및 판별 모델입니다. 기본적으로 생성 모델은 GPT에 몇가지 기사 정보를 조건부로 입력한 모델이고, 입력 시퀀스의 제일 마지막에 입력한 특수 토큰을 이용하여 discriminator를 학습한다고 합니다.
- 본 논문에서는 3가지 생성 설정을 실험에서 이용합니다.
    - **Model Sizes: Base (12 layers), Large (24 layers), Mega (48 layers)**
    - **Sampling Method:** top-k sampling, nucleus sampling
    - **Conditioning:** initial article conditioning. 모델에 입력하는 초기 문장의 길이

**Classifier Models**

- 각 모델 설정을 구별하기 위한 분류기 모델을 학습합니다. 기본적으로 각 모델 설정이 하나의 클래스라고 생각하고 multi-class classification 문제로 태스크를 치환해서 풉니다.
- 모델 종류
    - BoW(Linear)
        - simple bag-of-words 베이스라인.
        - 각 단어 임베딩을 평균하여 단일 선형 레이어 + 소프트맥스

            
         <img width="158" alt="스크린샷 2020-09-08 오후 7 24 39" src="https://user-images.githubusercontent.com/49383323/92465046-09b97d00-f209-11ea-8258-03d82d64e0ef.png">

    - BoW(MLP)
        - 선형 레이어 위에 ReLU 비선형 레이어 + 소프트맥스

            
      <img width="168" alt="스크린샷 2020-09-08 오후 7 24 44" src="https://user-images.githubusercontent.com/49383323/92465065-0f16c780-f209-11ea-9af9-1ad9255a9a2d.png">

    - ConvNet
        - 필터 3짜리 1D Convolution 레이어를 태우고 위에 선형 레이어 +소프트맥스
    - LSTM
        - LSTM 태우고 아웃풋을 mean-pool + Softmax
    - Transformer Encoders
        - 4 layer 짜리 multi-head Transformer 인코더

# 데이터

- News Corpora: CNN/Dailymail news corpus
    - 이 코퍼스는 총 90K 뉴스 기사로 구성되어 있습니다.
    - N개의 뉴스 기사를 시드 코퍼스로 주어주고, 각 모델 설정으로 N개의 뉴스 기사를 추가적으로 수집했습니다.
- 최대 시퀀스 길이를 500으로 제한하고, 데이터셋을 8:1:1 비율로 학습, 검증, 테스트로 나눴습니다.

# 실험

**실험 설정**

- 아래 표와 같이 10개의 태스크를 정의하여 생성된 문장으로부터 생성 모델의 모델 설정을 예측하는 태스크를 세웠습니다.

<img width="351" alt="스크린샷 2020-09-08 오후 7 26 06" src="https://user-images.githubusercontent.com/49383323/92465140-26ee4b80-f209-11ea-8625-c60d6d8a56ac.png">

- 또한, 사람이 직접 작성한 기사도 포함시킨 태스크도 실험해보았습니다.
- 단어 임베딩 차원을 64로 고정하고, 임베딩을 scratch부터 학습했습니다.
- 인코더의 히든 사이즈도 모두 64로 고정하고, 배치사이즈 64로 50에폭 돌렸습니다.
    - validation accuracy를 기준으로 patience 3으로 early stopping

**실험 결과**

- 아래의 결과를 보면, 랜덤으로 선택한 Chance 확률보다 대부분 더 높은 확률로 모델 설정을 예측할 수 있었습니다.

<img width="713" alt="스크린샷 2020-09-08 오후 7 26 37" src="https://user-images.githubusercontent.com/49383323/92465191-39688500-f209-11ea-9c46-31279ec02acb.png">

- 이 결과로부터, 특정 모델 설정이 어떠한 형태로든 잠재적인 특징을 갖고 있다는 사실을 알 수 있습니다.
- 또한, 각 모델 설정마다 예측 난이도가 다르다는 것을 알수 있습니다.
    - 예컨대, 모델 사이즈 (S1)을 예측하는 것이 nucleus p에 관한 설정보다 더 어려운 태스크라는 것을 알 수 있습니다.
    - 즉, 생성에 직접적으로 관련된 디코딩 method를 예측하는 것이 condition length나 모델 사이즈보다 어렵다고 해석할 수 있습니다.
- 또한, (P1,P2), (K1,K2) 결과를 보면, 서로 가까운 p or k 값을 구별하는 것이 서로 먼 p or k 값을 구분하는 태스크보다 더 어렵다는 사실을 알 수 있습니다.
    - 이것은 생성 문장이 어떠한 순서와 크기로 이동한다는 사실을 보여줍니다.
- 다양한 예측 모델에 대한 성능을 비교해보았을 때, Transformer가 단순 BoW Linear 레이어보다 성능이 떨어지는 경우가 종종 있습니다.
    - 이는, 문장에 내포된 특징이 compositional dependencies (단어 순서) 등이 아니라 style (단어 선택)에 더 크게 연관되어 있다는 것을 알려줍니다.
    - 다만, ConvNet이나 Transformer가 더 성능이 좋은 경우도 있는데, 이는 단어 앞뒤의 interaction이 간혹 유효하게 사용되는 경우가 있기 때문이라고 생각합니다.
    - LSTM은 대부분의 경우에서 제일 성능이 안좋았는데, complex compotional relations이 생성 문장에 내포된 특징을 포착하는데 크게 도움이 안된다는 사실을 알려줍니다.

<img width="740" alt="스크린샷 2020-09-08 오후 7 26 50" src="https://user-images.githubusercontent.com/49383323/92465210-3f5e6600-f209-11ea-8fd3-0017c35fe09f.png">

- 위의 표는 사람이 직접 작성한 기사도 추가하여 예측한 실험 결과입니다.
- 표 2와 비교해보았을 때 전체적으로 예측 성능이 향상되었습니다. 또한, 사람이 작성한 기사에 대해서는 90% 넘는 정확도로 모델이 예측한다는 사실을 발견했습니다.
- 이로부터, human-machine text를 구별하는 태스크가 machine-machine text를 구별하는 태스크보다 더 쉽다는 사실을 알 수 있습니다.
- 모델이 하이퍼파라미터를 조금 조정하는 정도로도 확연히 다른 문장을 생성한다는 것은 어찌보면, 지금 생성 모델들이 갖고 있는 한계라고 볼 수도 있습니다.

# 논의할만한 내용

- 예전에 본 논문 (THE CURIOUS CASE OF NEURAL TEXT DeGENERATION)에 있던 그림인데, Beam Search Text와 사람이 작성한 텍스트 사이에는 word choice 분포가 다르다는 것 나타내는 그림입니다.
    - 아마, Top-K, Nucleus를 쓴다고 해도 분명히 둘 사이에는 유의미한 word choice 분포 차이가 있을 것이기 때문에 human-machine text 구별이 가능하지 않을까 싶음.

![_2020-04-21__12 11 24](https://user-images.githubusercontent.com/49383323/92465248-4b4a2800-f209-11ea-85a2-43c14b62c459.png)

- 생성 모델이 생성한 문장이 최종적으로는 랭커에 입력될텐데 랭커가 생성 문장에 내포된 artifacts를 학습해버리는 부작용이 생길 수도 있지 않을까?
    - 지금 랭커는 word choice와 관련된 피쳐를 explicit하게 쓰고 있지 않기 때문에 아직은 그럴 일은 없다고 생각함
    - 그리고, 최대 500자짜리 기사보다 훨씬 짧으니까 이런 문제는 좀 덜 할 듯
- Vocab 사이즈나 토크나이저를 바꿔본 실험도 있었으면 더 좋았을듯

# 관련된 논문
